### ISSUE TITLE
Only free nodes can be locked

### ISSUE THREAD

**Issue by lpellegr** [ow2-proactive/scheduling#2748]
There are use cases for a customer where it should be possible to lock a node when it is busy (i.e. running a Task). This would allow waiting for the completion of the Task but prevent others executing.

Checking node state at client side is not an acceptable solution since compare and set will not be atomic, thus leading to race conditions.

As a workaround, it is possible to play with the state of the Scheduler: setting the Scheduler in frozen state will prevent Tasks and Jobs to be started but running ones to complete. However, this solution affects all Jobs and thus nodes. It is not possible for instance to prevent execution on a specific host for a maintenance that is planned.

In the code, there is a specific check performed for the state of the node before locking:

https://github.com/ow2-proactive/scheduling/blob/7.19.X/rm/rm-server/src/main/java/org/ow2/proactive/resourcemanager/core/RMCore.java#L1663

It seems that it is not a bug but at least a choice. I don't know if this choice was constrained by something. Deeper investigations would be required.

Personally, I think that locking should be allowed whatever the state of the node is.

**Comment by tmoulin** [comment]
As a workaround, it is possible to use the return value of the locking request and loop on it.
It is far from ideal though.

**Comment by lpellegr** [comment]
@tmoulin Unfortunately, this will not work in production depending of the requirements and nodes usage.

The reason lies in the fact that between each of the request you may perform for trying to lock a node (cf. your loop), nothing prevents the Scheduler to start a new Task on this node (which is a common use case if the Scheduler is heavily loaded: has several pending Jobs). In such a situation, the request to lock the node will fail again and again (starvation).

Hereafter is a use case example. I have a customer who is executing on a ProActive node Jobs, each having a single Task that take 2 hours to complete. While the Task is executing, he suddenly detects that a software maintenance is required on the machine hosting the ProActive node and he would like to start this maintenance as soon as the running Task complete. The first idea idea would be to lock the ProActive node, thus allowing the Task to complete and prevent new ones to start. Unfortunately, this is not possible to lock a busy node for now. As a consequence, the customer can try what you suggested, in a loop. However, the Scheduler has 3 pending Jobs ready to execute on that node. Since the Scheduler scheduling loops tries every 100 milliseconds (default value) to schedule a new Task on nodes, it may happen that a new Task is scheduled for execution between 2 of the requests made to lock the nodes. This means that in an unlucky situation (which happens more often than we think) the customer has to wait 6 additional hours before starting the maintenance.

**Comment by lpellegr** [comment]
It is currently not possible to lock a node that is not free. A simple solution is to update the Scheduler core to allow locking whatever the state is. This change seems to work but it reveals a design issue. 

When a node is locked, its state change to _LOCKED_. It means that it is no longer possible to know if the node is locked but busy for Task completion, or free, etc. Another consequence, it that applying the previous simple solution will make a node busy but locked appear green in the RM Portal and other clients, which is really misleading.

My suggestion is to update the code so that:

  1. A node can be locked whatever the state it has
  2. It is possible to distinguish the internal state managed by the Scheduler from the locked status

Regarding the second suggestion, there are at least 2 possibilities: 

  - Removing LOCK state from `NodeState` and duplicating all remaining states with new ones having the suffix  _LOCKED_.
  - Removing LOCK state from `NodeState` and keep this information as an additional attribute in `RMNodeImpl`

I prefer the second possibility since it avoids to check and manage transitions between all states but also clearly keep separated internal states managed by the RM from statuses defined by manual actions.

One last thing related to the locking of nodes. I noticed that a ProActive node marked as locked is automatically unlocked on Scheduler restart. From my point of view and based on the use case I had with one customer, it is an issue. 

@bamedro Is it a normal behaviour or can it be considered as an issue to address?

**Comment by lpellegr** [comment]
I have discussed with Brian. He agrees that removing ProActive node locks on Scheduler restart is problematic. I will create a dedicated issue.
